File: core/src/main/scala/dev/akif/e/DefaultEncoderE.scala
package dev.akif.e

import scala.collection.JavaConverters.mapAsScalaMap

object DefaultEncoderE extends EncoderE[String] {
  override def encode(e: E): String = {
    val strings = List(
      s""""code":${e.code}""",
      if (!e.hasName)    "" else s""""name":"${escape(e.name)}"""",
      if (!e.hasMessage) "" else s""""message":"${escape(e.message)}"""",
      if (!e.hasCause)   "" else s""""cause":"${escape(e.cause.getMessage)}"""",
      if (!e.hasData)    "" else s""""data":${mapAsScalaMap(e.data).map({case (k, v) => s""""${escape(k)}":"${escape(v)}""""}).mkString("{", ",", "}")}"""
    )

    val size = strings.size

    val (_, builder) = strings.foldLeft(1 -> new StringBuilder("{")) {
      case ((i, sb), "") if i == size && sb.last == ',' => (i + 1) -> sb.dropRight(1)
      case ((i, sb), "")                                => (i + 1) -> sb
      case ((i, sb), s) if i < size                     => (i + 1) -> sb.append(s).append(",")
      case ((i, sb), s)                                 => (i + 1) -> sb.append(s)
    }

    builder.append("}").toString()
  }

  private def escape(s: String): String = s.replaceAll("\"", "\\\\\"")
}

File: core/src/main/java/dev/akif/e/DefaultEncoderE.java
package dev.akif.e;

import java.util.Map;
import java.util.StringJoiner;

public final class DefaultEncoderE implements EncoderE<String> {
    private static DefaultEncoderE instance;

    private DefaultEncoderE() {}

    public static DefaultEncoderE get() {
        if (instance ==  null) {
            instance = new DefaultEncoderE();
        }

        return instance;
    }

    @Override public String encode(E e) {
        StringJoiner joiner = new StringJoiner(",", "{", "}");

        if (e.hasCode())    { joiner.add(String.format("\"code\":%d",        e.code)); }
        if (e.hasName())    { joiner.add(String.format("\"name\":\"%s\"",    escape(e.name))); }
        if (e.hasMessage()) { joiner.add(String.format("\"message\":\"%s\"", escape(e.message))); }
        if (e.hasCause())   { joiner.add(String.format("\"cause\":\"%s\"",   escape(e.cause.getMessage()))); }
        if (e.hasData())    { joiner.add(String.format("\"data\":%s",        makeDataString(e.data))); }

        return joiner.toString();
    }

    String escape(String s) {
        return s.replaceAll("\"", "\\\\\"");
    }

    String makeDataString(Map<String, String> data) {
        StringJoiner joiner = new StringJoiner(",", "{", "}");
        data.forEach((key, value) -> joiner.add(String.format("\"%s\":\"%s\"", escape(key), escape(value))));
        return joiner.toString();
    }
}
